presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/BaseJdbcClient.java

```diff
@@ -141,38 +200,198 @@ public BaseJdbcClient(JdbcConnectorId connectorId, BaseJdbcConfig config, String
     }
 
     @Override
+    public Set<String> getSchemaNames()
+    {
+        ImmutableSet.Builder<String> schemaNames = ImmutableSet.builder();
+        Set<String> originalNames = getOriginalSchemas();
+        Map<String, Optional<String>> mappedNames = new HashMap<>();
+        for (String schemaName : originalNames) {
+            String schemaNameLower = schemaName.toLowerCase(ENGLISH);
+            schemaNames.add(schemaNameLower);
+            mappedNames.put(schemaNameLower, Optional.of(schemaName));
+        }
+
+        // if someone is listing all of the schema names, throw them all into the cache as a refresh since we already spent the time pulling them from the DB
+        schemaMappingCache.putAll(mappedNames);
+        return schemaNames.build();
+    }
+
+    /**
+     * Pull the list of Table names from the RDBMS exactly as they are returned. Override if the RDBMS method to pull the tables is different. Each {@link String} array must have
+     * the table's original schema name as the first element, and the original table name as the second element.
+     *
+     * @param schema the schema to list the tables for, or NULL for all
+     * @return the schema + table names
+     */
+    protected List<CaseSensitiveMappedSchemaTableName> getOriginalTablesWithSchema(Connection connection, String schema)
+    {
+        try (ResultSet resultSet = getTables(connection, schema, null)) {
+            ImmutableList.Builder<CaseSensitiveMappedSchemaTableName> list = ImmutableList.builder();
+            while (resultSet.next()) {
+                String schemaName = resultSet.getString("TABLE_SCHEM");
+                String tableName = resultSet.getString("TABLE_NAME");
+                list.add(new CaseSensitiveMappedSchemaTableName(schemaName, tableName));
+            }
+            return list.build();
+        }
+        catch (SQLException e) {
+            throw new PrestoException(JDBC_ERROR, e);
+        }
+    }
+
+    @Override
     public List<SchemaTableName> getTableNames(@Nullable String schema)
     {
         try (Connection connection = driver.connect(connectionUrl, connectionProperties)) {
             DatabaseMetaData metadata = connection.getMetaData();
-            if (metadata.storesUpperCaseIdentifiers() && (schema != null)) {
-                schema = schema.toUpperCase(ENGLISH);
+
+            schema = finalizeSchemaName(metadata, schema);
+            Map<String, Map<String, String>> schemaMappedNames = new HashMap<>();
+            ImmutableList.Builder<SchemaTableName> list = ImmutableList.builder();
```

Comment:
call this tableNames, perhaps?

Reply:
You mean the list? It's a local only variable, but sure.